aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -170,8 +183,17 @@ SparseTensor& div_out_sparse_zerodim(SparseTensor& r, const SparseTensor& t, con
   AT_ASSERT(t.is_sparse());
 
   if (is_same_tensor(r, t)) {
+    // Can't divide an uncoalesced integral tensor accurately. e.g. for a sparse int tensor with value 6
+    // represented as values=[3, 3], integer division by 2 would give values=[1, 1] => 2 instead
+    // of 6 / 2 => 3
+    if (!r.is_coalesced() && isIntegralType(r.scalar_type(), true)) {
+      coalesce_(r);
+    }
     r._values().div_(value);
   } else {
+    if (!t.is_coalesced() && isIntegralType(r.scalar_type(), true)) {
+      coalesce_(t);
```

Comment:
t is marked as a `const SparseTensor&`, which means that we shouldn't be modifying it in-place.

Reply:
is this a general C++ convention / best practice for readability, or does modification of attributes of a const reference cause potential issues?

I was thinking of `const` as being info for the compiler, not as being an annotation for readers, and assumed that since this modification was allowed that it's fine. It's confusing to me because in JVM languages it's considered good practice to make a value "final" (or "val") even if the object it refers to is mutable.

Since the semantics of the data held in `t` don't change, doing the coalesce out of place seems wasteful (the caller may eventually need to coalesce `t` again), but I don't think I have the option of changing t to be non-const.

Reply:
Const correctness is using the keyword const to prevent `const` objects from being mutated. Const references should not be modified; when they do get modified the compiler should complain about it. Here is an overview of const correctness: https://isocpp.org/wiki/faq/const-correctness#overview-const .

It's both a best practice for readability and an annotation for the compiler: as a reader, I expect a const reference to not get modified and I expect the compiler to enforce it. Any deviation from those two is fishy. In this case, we've deviated from both of those points: we're modifying a const reference, AND the compiler isn't enforcing it!

Now, the reason why the compiler doesn't complain in this case is that `tensor._values()` returns a Tensor and one can arbitrarily mutate a Tensor. We are not implementing this correctly (but I haven't thought about what exactly needs to be done here).

Fixing `_values()` such that we catch cases like this is beyond the scope of this PR, but you're welcome to do it. Fixing the code to make it not look like a red flag to readers is important and should be done in this PR.

Reply:
> I was thinking of `const` as being info for the compiler, not as being an annotation for readers, and assumed that since this modification was allowed that it's fine. It's confusing to me because in JVM languages it's considered good practice to make a value "final" (or "val") even if the object it refers to is mutable.

I'm not familiar with the `final` keyword so I can't make a comparison here

> Since the semantics of the data held in `t` don't change, doing the coalesce out of place seems wasteful (the caller may eventually need to coalesce `t` again), but I don't think I have the option of changing t to be non-const.

Let's think about it from another perspective. Let's say the user is doing a division of an (integer) sparse tensor with an integer scalar. After this division, their original tensor got modified! 

There's an argument to made that coalescing a tensor in-place is not actually a modification, especially if we're trying to hide coalesced-ness from the user.

Reply:
> I'm not familiar with the final keyword so I can't make a comparison here

From a language perspective, C++ const and java final variables are basically the same afaik. In java, `final` variables just can't be re-assigned (the meaning of final java methods vs const c++ functions is different though). Similar language semantics doesn't mean coding conventions / best practices are the same though, and I could conceive of C++ compilers making optimizations that rely on attributes of const objects not having been modified (if the language standard says that's allowed?).

It seems like if the language wanted to allow optimizations based on this, they wouldn't allow one to call non-const functions on non-mutable attributes of a const reference though, which they do allow (since we can call functions like `copy_` on `indices_`/`values_`).

> as a reader, I expect a const reference to not get modified and I expect the compiler to enforce it.

I think this might be a difference in convention. When you say "it" though we're referring specifically to attributes of the object referred to, not to the const reference itself.

> Now, the reason why the compiler doesn't complain...

ahh, interesting. I hadn't paid much attention to the constness of tensor functions. Makes sense.

> We are not implementing this correctly ...

These are considered "unsafe"/raw access so could be intentional? See the "note" on `uncoalesced` here:
https://pytorch.org/docs/stable/sparse.html

Is there a way to restrict calls to non-const functions on attributes of a const class?

> There's an argument to made that coalescing a tensor in-place is not actually a modification, especially if we're trying to hide coalesced-ness from the user.

Right, and I'm also questioning that, a use case described here:

https://isocpp.org/wiki/faq/const-correctness#mutable-data-members

The coalesce benefits the caller, and doesn't change the values of the tensor. It's sort of like caching -- I would expect that an object containing a cache could update its cache even if referred to by a const reference. In a sense a coalesce is a cache of what will eventually need to be materialized. We do expose `is_coalesced()`, but that seems like an implementation detail similar to if someone exposed an `is_cached()` on an object that might cache the result of internal computations on its attributes. 

From the note referenced above, the user isn't supposed to care about the internal representation and just consider it an optimization detail.

In terms of consistency though, I do see that functions like `set_coalesced` is not marked as const. We could have hypothetically chose to have `coalesced` be marked as `mutable` in SparseTensor and didn't, so that makes me more partial to the argument that we should take the performance hit and do this out of place here.